serviceregistration/kubernetes/retry_handler_test.go

```diff
@@ -0,0 +1,243 @@
+		t.Fatal(err)
+	}
+	if len(r.patchesToRetry) != 1 {
+		t.Fatal("expected 1 patch")
+	}
+
+	if err := r.Add(testPatch2); err != nil {
+		t.Fatal(err)
+	}
+	if len(r.patchesToRetry) != 2 {
+		t.Fatal("expected 2 patches")
+	}
+
+	if err := r.Add(testPatch3); err != nil {
+		t.Fatal(err)
+	}
+	if len(r.patchesToRetry) != 3 {
+		t.Fatal("expected 3 patches")
+	}
+
+	if err := r.Add(testPatch4); err != nil {
+		t.Fatal(err)
+	}
+	if len(r.patchesToRetry) != 4 {
+		t.Fatal("expected 4 patches")
+	}
+
+	// Adding a dupe should result in no change.
+	if err := r.Add(testPatch4); err != nil {
+		t.Fatal(err)
+	}
+	if len(r.patchesToRetry) != 4 {
+		t.Fatal("expected 4 patches")
+	}
+
+	// Adding a reversion should result in its twin being subtracted.
+	if err := r.Add(&client.Patch{
+		Operation: client.Add,
+		Path:      "four",
+		Value:     "false",
+	}); err != nil {
+		t.Fatal(err)
+	}
+	if len(r.patchesToRetry) != 3 {
+		t.Fatal("expected 3 patches")
+	}
+
+	if err := r.Add(&client.Patch{
+		Operation: client.Add,
+		Path:      "three",
+		Value:     "false",
+	}); err != nil {
+		t.Fatal(err)
+	}
+	if len(r.patchesToRetry) != 2 {
+		t.Fatal("expected 2 patches")
+	}
+
+	if err := r.Add(&client.Patch{
+		Operation: client.Add,
+		Path:      "two",
+		Value:     "false",
+	}); err != nil {
+		t.Fatal(err)
+	}
+	if len(r.patchesToRetry) != 1 {
+		t.Fatal("expected 1 patches")
+	}
+
+	if err := r.Add(&client.Patch{
+		Operation: client.Add,
+		Path:      "one",
+		Value:     "false",
+	}); err != nil {
+		t.Fatal(err)
+	}
+	if len(r.patchesToRetry) != 0 {
+		t.Fatal("expected 0 patches")
+	}
+}
+
+// This is meant to be run with the -race flag on.
+func TestRetryHandlerRacesAndDeadlocks(t *testing.T) {
+	_, testConf, closeFunc := kubetest.Server(t)
+	defer closeFunc()
+
+	client.Scheme = testConf.ClientScheme
+	client.TokenFile = testConf.PathToTokenFile
+	client.RootCAFile = testConf.PathToRootCAFile
+	if err := os.Setenv(client.EnvVarKubernetesServiceHost, testConf.ServiceHost); err != nil {
+		t.Fatal(err)
+	}
+	if err := os.Setenv(client.EnvVarKubernetesServicePort, testConf.ServicePort); err != nil {
+		t.Fatal(err)
+	}
+
+	logger := hclog.NewNullLogger()
+	shutdownCh := make(chan struct{})
+	wait := &sync.WaitGroup{}
+	testPatch := &client.Patch{
+		Operation: client.Add,
+		Path:      "patch-path",
+		Value:     "true",
+	}
+
+	c, err := client.New(logger, shutdownCh)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	r := &retryHandler{
+		logger:         logger,
+		namespace:      kubetest.ExpectedNamespace,
+		podName:        kubetest.ExpectedPodName,
+		client:         c,
+		patchesToRetry: make([]*client.Patch, 0),
+	}
+	go r.Run(shutdownCh, wait)
+
+	// Now hit it as quickly as possible to see if we can produce
+	// races or deadlocks.
+	start := make(chan struct{})
+	done := make(chan bool)
+	numRoutines := 100
+	for i := 0; i < numRoutines; i++ {
+		go func() {
+			<-start
+			if err := r.Add(testPatch); err != nil {
+				t.Fatal(err)
```

Comment:
`t.FailNow` (which `t.Fatal` calls) should only be called from the main goroutine.  https://golang.org/pkg/testing/#T.FailNow

Reply:
Oh, I didn't know that! :-) Thanks!

Reply:
https://github.com/hashicorp/vault/pull/8249/commits/a89bc1f228fb5780247caeafdb411a3620895036